framework/src/functormaterials/ParsedFunctorMaterial.C

```diff
@@ -0,0 +1,134 @@
+
+#include "ParsedFunctorMaterial.h"
+
+registerMooseObject("MooseApp", ParsedFunctorMaterial);
+registerMooseObject("MooseApp", ADParsedFunctorMaterial);
+
+template <bool is_ad>
+InputParameters
+ParsedFunctorMaterialTempl<is_ad>::validParams()
+{
+  InputParameters params = FunctorMaterial::validParams();
+  params += FunctionParserUtils<is_ad>::validParams();
+
+  params.addClassDescription(
+      "Computes a functor material from a parsed expression of other functors.");
+
+  params.addRequiredCustomTypeParam<std::string>(
+      "expression", "FunctionExpression", "Expression to parse for the new functor material");
+  params.addParam<std::vector<std::string>>("functor_names",
+                                            "Functors to use in the parsed expression");
+  params.addParam<std::vector<std::string>>(
+      "functor_symbols",
+      "Symbolic name to use for each functor in 'functor_names' in the parsed expression. If not "
+      "provided, then the actual functor names must be used in the parsed expression.");
+  params.addRequiredParam<std::string>("property_name",
+                                       "Name to give the new functor material property");
+
+  return params;
+}
+
+template <bool is_ad>
+ParsedFunctorMaterialTempl<is_ad>::ParsedFunctorMaterialTempl(const InputParameters & parameters)
+  : FunctorMaterial(parameters),
+    FunctionParserUtils<is_ad>(parameters),
+    _expression(getParam<std::string>("expression")),
+    _functor_names(getParam<std::vector<std::string>>("functor_names")),
+    _n_functors(_functor_names.size()),
+    _functor_symbols(getParam<std::vector<std::string>>("functor_symbols")),
+    _property_name(getParam<std::string>("property_name"))
+{
+  // Check/modify 'functor_symbols'
+  if (_functor_symbols.size() != _n_functors)
+  {
+    if (_functor_symbols.size() == 0)
+      _functor_symbols = _functor_names;
+    else
+      paramError("functor_symbols",
+                 "The number of entries must be equal to either zero or the number of entries in "
+                 "'functor_names'.");
+  }
+
+  // Get the functors
+  _functors.resize(_n_functors);
+  for (const auto i : index_range(_functor_names))
+    _functors[i] = &getFunctor<GenericReal<is_ad>>(_functor_names[i]);
+
+  // Build the parsed function
+  buildParsedFunction();
+
+  // Add the functor material property
+  addFunctorProperty<GenericReal<is_ad>>(
+      _property_name,
+      [this](const auto & r, const auto & t) -> GenericReal<is_ad>
+      {
+        // Store the functor values
+        for (const auto i : index_range(_functors))
+          _func_params[i] = (*_functors[i])(r, t);
+
+        // Store the space and time values
+        const auto r_point = r.getPoint();
+        _func_params[_n_functors] = r_point(0);
+#if LIBMESH_DIM > 1
+        _func_params[_n_functors + 1] = r_point(1);
+#endif
+#if LIBMESH_DIM > 2
+        _func_params[_n_functors + 2] = r_point(2);
+#endif
```

Comment:
I think we can safely discard the use of these macros here. It hinders readability. If we don't configure libMesh with a dimension of 3, I'd be shocked if a lot of things in MOOSE didn't die horribly

Reply:
Works for me